File: c++/1629.cpp
#include <iostream>

using namespace std;

long long A, B, C;

long long solve(long long power) {
  if (power == 0) return 1;
  if (power == 1) return A % C;

  long long half = solve(power / 2) % C;

  if (power % 2 == 1)
    return half * half % C * A % C;
  else
    return (half * half) % C;
}

int main() {
  freopen("example.txt", "r", stdin);
  cin >> A >> B >> C;

  cout << solve(B);
  return 0;
}

File: js/1629.js
const fs = require('fs');

const [A, B, C] = fs
  .readFileSync('example.txt')
  .toString()
  .trim()
  .split(' ')
  .map(BigInt);

console.log(solve(B).toString());

function solve(power) {
  if (power === 1n) return A % C;

  const half = solve(power / 2n) % C;

  if (power % 2n === 1n) return (half * half * (A % C)) % C;
  else return (half * half) % C;
}
